src/metabase/api/card.clj

```diff
@@ -280,24 +267,58 @@
+        metadata-timeout     (a/timeout metadata-sync-wait-ms)
+        [metadata port]      (a/alts!! [result-metadata-chan metadata-timeout])
+        timed-out?           (= port metadata-timeout)
+        card                 (db/transaction
+                              ;; Adding a new card at `collection_position` could cause other cards in this
+                              ;; collection to change position, check that and fix it if needed
+                              (api/maybe-reconcile-collection-position! card-data)
+                              (db/insert! Card (cond-> card-data
+                                                 (not timed-out?)
+                                                 (assoc :result_metadata metadata))))]
+    (events/publish-event! :card-create card)
+    (log/info (trs "Metadata not available soon enough. Saving new card and asynchronously updating metadata"))
+    ;; include same information returned by GET /api/card/:id since frontend replaces the Card it currently has with
+    ;; returned one -- See #4283
+    (u/prog1 (-> card
+                 (hydrate :creator
+                          :dashboard_count
+                          :can_write
+                          :average_query_time
+                          :last_query_start
+                          :collection [:moderation_reviews :moderator_details])
+                 (assoc :last-edit-info (last-edit/edit-information-for-user @api/*current-user*)))
+      (when timed-out?
+        (.submit clojure.lang.Agent/pooledExecutor
+                 ^Runnable
+                 (fn []
+                   (try
+                     (let [timeoutc        (a/timeout metadata-async-timeout)
+                           [metadata port] (a/alts!! [result-metadata-chan timeoutc])
+                           id              (:id <>)
+                           current-query   (db/select-one-field :dataset_query Card :id id)]
+                       (cond (= port timeoutc)
+                             (log/info (trs "Metadata not ready in 15 minutes, abandoning"))
+
+                             (and (seq metadata) (= current-query (:dataset_query <>)))
+                             (do (db/update! Card id {:result_metadata metadata})
+                                 (log/info (trs "Metadata updated asynchronously for card {0}" id)))
+
+                             (not= current-query (:dataset_query <>))
+                             (log/info (trs "Not updating metadata asynchronously for card {0} because query has changed"
+                                            id))
+
+                             (not (seq metadata))
+                             (log/info (trs "Not updating metadata asynchronously for card {0} because no metadata"
+                                            id))))
+                     (catch Exception e
+                       (log/warn e (trs "Error updating card metadata asynchronously"))))))))))
```

Comment:
This is a mega function (66 lines) -- can any parts of it be split out into separate functions? It's a bit hard to read huge functions. But maybe the async metadata update stuff can be split out